### DISCUSSION TITLE
Can Ocelot have own Controllers & Actions?

### DISCUSSION THREAD

**Discussion by kvansaders** [ThreeMammals/Ocelot#1949] [Q&A] [answered]
## Trying to create API gateway which also uses it's own Web API Controllers to handle mashups and any other custom code, using Ocelot for pipeline items like rate limiting, Authorization, and caching.

I am trying to put together an API gateway that also serves the purpose of mashing up calls to multiple APIs. I would also like to have autogenerated Swagger, so I'd like to combine the capability of Ocelot for authorization and rate limiting but would also like to be able to create the edge custom API's in .NET code to handle translations and combinations of API calls.

Is this possible with Ocelot, and are there any examples? Does Ocelot only work with downstream API's as separate applications? I'm trying to avoid an extra hop that would only exist for mashups and composite API calls.

Stack Overflow related item: https://stackoverflow.com/questions/56956831/can-ocelot-have-its-own-controller-actions

**Comment by danielkaroldudek** [comment, marked as answer]
Hi @kvansaders,
Recently I was working on similar PoC for one of the projects I'm working on and my main goal was to have endpoints like @kvansaders mentioned to be able to have possibility mehsup / service enhance / add specific aggregation in case of one call to API G. with multiple walls internally or hide old / ugly API and what was more important for me  to be able to call to API G. to publish a message to a broker. My first approach was to have middleware which worked but I wasn't able to create out of box swagger. Then I get to this topic of Ocelot own APIs.
After standard Ocelot configuration on .NET 6 project with some downstream projects what I did additionally is:
1) In the services section (Program.cs) make sure you have `services.AddEndpointsApiExplorer();`
and if you would like to have swagger use MMLib.SwaggerForOcelot nugget with configuration
```csharp
services.AddSwaggerForOcelot(configuration, (o) =>
    {
        o.GenerateDocsForGatewayItSelf = true;
    });
```
2) Then in the app section (Program.cs) for swagger make sure you have `app.UseSwaggerForOcelotUI();`
and your custom Api like `app.MapGet("/testing", () => Results.Ok("It's working"));`
This way swagger will start generating the doc (change the view from default first service to Gateway - there you will see your endpoint). In my case running this I was getting 404 when I called the internal API. After couple of tryouts with the configuration I figured out that I was missing:
```csharp
app.UseRouting();
app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers();
});
```
With such configuration you will be able to run standard downstream / upstream setup with ocelot and swagger (need additional configuration in ocelot.json) and your custom api for any needs.
If needed let me know I may publish some demo app into the github. I hope this will help you guys.

**Reply by raman-m** [reply]
@kvansaders 
Here is your answer! Enjoy the solution!

**Comment by berkslv** [comment]
I'm sharing the entire Program.cs code if anyone is interested. Unfortunately [ASP0014](https://learn.microsoft.com/en-us/aspnet/core/diagnostics/asp0014?view=aspnetcore-8.0) error cannot be suprased without `#pragma warning disable` statement. The `app.MapControllers()` does not work.

```csharp

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddOcelot(builder.Configuration);

var app = builder.Build();

app.UseRouting();
app.MapControllers();
#pragma warning disable ASP0014
app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers();
});
#pragma warning restore ASP0014

app.UseOcelot();

app.MapGet("/identity/authenticate", () => "token");

app.Run();

```

**Reply by raman-m** [reply]
What will happen if to define a route with upstream path `"/identity/authenticate` ? 😉

**Reply by cjbush** [reply]
> What will happen if to define a route with upstream path `"/identity/authenticate` ? 😉

If you do that, you get what you deserve. It's not a weird use case to want to be able to have custom endpoints in an API gateway though.
